perfdash/perfdash.go

```diff
@@ -134,37 +140,94 @@ func parseTestOutput(scanner *bufio.Scanner, buildNumber int, resources sets.Str
 	return hist
 }
 
-func getLatencyData(client *jenkins.JenkinsClient, job string) (TestToBuildData, sets.String, sets.String, error) {
+// returns build number of the next run greater than *startFrom
+// TODO: if anyone has any idea how to do it better, please do
+func binSearch() int {
+	start := *startFrom
+	end := *startFrom + 1
+	// search from some build number without result
+	response, err := http.Get(fmt.Sprintf("%v%v%v", urlPrefix, start, urlSuffix))
+	if err != nil {
+		panic(err)
+	}
+	if response.StatusCode != 200 {
+		return start
+	}
+	response, err = http.Get(fmt.Sprintf("%v%v%v", urlPrefix, end, urlSuffix))
+	if err != nil {
+		panic(err)
+	}
+	if response.StatusCode != 200 {
+		return end
+	}
+	inc := 1
+	for response.StatusCode == 200 {
+		inc *= 2
+		start = end
+		end += inc
+		response, err = http.Get(fmt.Sprintf("%v%v%v", urlPrefix, end, urlSuffix))
+		if err != nil {
+			panic(err)
+		}
+	}
+	// We know that data exists for start and does not exist for end. Now binary search.
+	for start+1 < end {
```

Comment:
I think it would be simpler to do:

for start < end {
  candidate = (start + end) / 2
  if response == 200 {
    start = candidate+1
  } else {
    end = candidate 
  }
}

Then you are guaranteed to have start=end at the end, and don't need the rest of the code.